src/passes/Inlining.cpp

```diff
@@ -143,22 +160,18 @@ struct FunctionInfo {
     if (options.shrinkLevel > 0 || options.optimizeLevel < 3) {
       return false;
     }
-    if (hasCalls) {
-      // This has calls. If it is just a trivial call itself then inline, as we
-      // will save a call that way - basically we skip a trampoline in the
-      // middle - but if it is something more complex, leave it alone, as we may
-      // not help much (and with recursion we may end up with a wasteful
-      // increase in code size).
-      //
-      // Note that inlining trivial calls may increase code size, e.g. if they
-      // use a parameter more than once (forcing us after inlining to save that
-      // value to a local, etc.), but here we are optimizing for speed and not
-      // size, so we risk it.
-      return isTrivialCall;
-    }
-    // This doesn't have calls. Inline if loops do not prevent us (normally, a
-    // loop suggests a lot of work and so inlining is less useful).
-    return !hasLoops || options.inlining.allowFunctionsWithLoops;
+    // The function just calls another function, but it's using locals in
+    // different order than the argument order, and/or using some locals more
+    // than once. In this case we inline if we're not optimizing for code size,
+    // as inlining it to more than one call site may increase code size by
+    // introducing locals.
+    if (trivialCall == TrivialCall::MayNotShrink) {
+      return true;
+    }
+    // Trivial calls are already handled. Inline if
+    // 1. The function doesn't have calls, and
+    // 2. The function doesn't have loops, or we allow inlining with loops.
```

Comment:
I reworded the old comment block here but it just repeats the single line of code below, so I think we can also just drop this comment.